Fix bug with image.scale 'bicubic' and ByteTensors.

jagapiou opened the pull request:
Bicubic interpolation was not working for scaling torch.ByteTensors.

* Temporary variables were also uint8, made them doubles.
* Added unit tests for ByteTensors.

soumith commented:
cc: @deltheil 
i wont get to review it for a few days (conference deadlines). if you have time, would be good :)

deltheil commented:
My main concern is related to the computation of the [interpolated value](https://github.com/torch/image/pull/118/files#diff-34b662b12392947cb0b99c2b3c91e9f6R170):

```C
dst[dst_pos] = a0 + si_f * (a1 + si_f * (a2 + a3 * si_f));
```

If the interpolated value gets negative or higher than 255 then it overflows when stored as a byte (`unsigned char`) on `dst`, .e.g. toy example:

```Lua
require "image"
local img = torch.ByteTensor{{0, 1, 32}}
local out = image.scale(img, 4, 1, 'bicubic')
-- This gives [0, 255, 9, 32] (the 2nd value overflows)
-- With float-s we would have [0.0000, -1.5556, 9.1111, 32.0000]
```

*(this is known as [overshooting](https://en.wikipedia.org/wiki/Bicubic_interpolation#Use_in_computer_graphics))*

So should we add some clamping on the interpolated value before storing it? (at least when working with bytes)

What when working with float-s/double-s? (undershoot)

Is it really that important to perform the scale computations from bytes vs. doing it with float-s + clamp / convert to byte a posteriori?

Can you think about a good strategy / spec here?

*Apart from that I was also wondering if there is a possibility to re-use / put in common the bicubic interpolation used by image warp: see [here](https://github.com/torch/image/blob/d97d20d/generic/image.c#L1640-L1680). But it's minor.*

deltheil commented on test/test_scale.lua:
```diff
@@ -14,35 +14,65 @@ local function outerProduct(x)
+    local areEqual = torch.eq(actual, expected):all()
+    tester:assert(areEqual)
+  else
+     print(torch.type(expected))
```
This should be removed, right?

jagapiou replied:
Done.

jagapiou commented:
Had a look at `image_(Main_bicubicInterpolate)` but that's finding the interpolated value for a single point, so the current implementation is more efficient I think. However, I have updated the code to use `image_(Main_cubicInterpolate)` to minimize code reuse. Note this function also suffered from the same under/overflow issues (now fixed). I think similar issues may be lurking elsewhere...

deltheil commented:
@jagapiou thanks! Looks better like that IMHO (cc @soumith @colesbury)

Could you squash your commits into a [single one](https://github.com/torch/torch7/blob/master/CONTRIBUTING.md)?

> I think similar issues may be lurking elsewhere

True...

jagapiou commented:
Done. Thanks for the quick response.

soumith commented:
Thanks a lot John!
